### PR TITLE
[FEATURE] TimeSeriesChartPanel can show values in legend

### PR DESCRIPTION
**Pull request by juliepagano**
# Notes for reviewers

- This PR follows on from https://github.com/perses/perses/pull/1230
- There are a few additional tasks that I don't want to do in this PR in the name of small PRs and making incremental progress on delivering the table legend. I've noted these in TODOs in the code. Let me know if you feel strongly about any of them getting done now instead of in follow-ups.
- I'm holding off on updating the migpart migration scripts until this feature is a little farther along, but they are on my radar to do in a follow-up PR. 

# Checklist

- [x] Pull request has a descriptive title and context useful to a reviewer.
- [x] Pull request title follows the `[<catalog_entry>] <commit message>` naming convention using one of the following `catalog_entry` values: `FEATURE`, `ENHANCEMENT`, `BUGFIX`, `BREAKINGCHANGE`, `IGNORE`.
- [x] All commits have [DCO signoffs](https://github.com/probot/dco#how-it-works).

## UI Changes

- [x] Changes that impact the UI include screenshots and/or screencasts of the relevant changes.
- [x] Code follows the [UI guidelines](https://github.com/perses/perses/blob/main/ui/ui-guidelines.md).
- [x] Visual tests are stable and unlikely to be flaky. See [Storybook](https://github.com/perses/perses/tree/main/ui/storybook#visual-tests) and [e2e](https://github.com/perses/perses/tree/main/ui/e2e#visual-tests) docs for more details. Common issues include:
  - Is the data inconsistent? You need to mock API requests.
  - Does the time change? You need to use consistent time values or mock time utilities.
  - Does it have loading states? You need to wait for loading to complete.

# Screenshots

image

image

### CODE REVIEW

File: ui/components/src/Drawer/Drawer.tsx [hunk 1]
```diff
@@ -20,7 +20,7 @@ export interface DrawerProps extends MuiDrawerProps {
   anchor?: 'left' | 'right';
 }
 
-const DRAWER_DEFAULT_WIDTH = 900;
+const DRAWER_DEFAULT_WIDTH = 1080;
```
Comment (juliepagano) [inline ui/components/src/Drawer/Drawer.tsx:23]: Making this a little wider to match updated designs and leave a little more breathing room for values in the chart view in the panel editor.

File: ui/components/src/OptionsEditorLayout/OptionsEditorControl.tsx [hunk 2]
```diff
@@ -52,7 +52,7 @@ export const OptionsEditorControl = ({ label, control, description }: OptionsEdi
             </InfoTooltip>
           )}
         </Stack>
-        <Box sx={{ width: '150px', textAlign: 'right' }}> {React.cloneElement(control, controlProps)}</Box>
+        <Box sx={{ width: '180px', textAlign: 'right' }}> {React.cloneElement(control, controlProps)}</Box>
```
Comment (juliepagano) [inline ui/components/src/OptionsEditorLayout/OptionsEditorControl.tsx:55]: Increasing this to give more space to the dropdowns now that the panel editor is wider.

File: ui/panels-plugin/src/plugins/time-series-chart/TimeSeriesChartPanel.tsx [hunk 3]
```diff
@@ -217,6 +236,36 @@ export function TimeSeriesChartPanel(props: TimeSeriesChartProps) {
     };
   }, [queryResults, thresholds, selectedLegendItems, legend, visual, isFetching, isLoading, y_axis?.max, y_axis?.min]);
```
Comment (juliepagano) [inline ui/panels-plugin/src/plugins/time-series-chart/TimeSeriesChartPanel.tsx:238]: I'm not in love with putting another large chunk of code into `TimeSeriesChartPanel`, but I'm a little burned out on yak shaving to decompose this component at this point and really need to get the table legend up-and-running, so I'm doing it. Let me know if you feel strongly about not doing so.

File: ui/panels-plugin/src/plugins/time-series-chart/TimeSeriesChartPanel.tsx [hunk 4]
```diff
@@ -217,6 +236,37 @@ export function TimeSeriesChartPanel(props: TimeSeriesChartProps) {
...
+    return legendValues.reduce((columns, value) => {
+      if (legendValues.includes(value)) {
+        columns.push({
+          accessorKey: `data.${value}`,
+          header: LEGEND_VALUE_CONFIG[value].label,
+          // TODO: evaluate making these dynamic to better account for the variability
+          // in space that numbers can take up. This requires revisiting the approach
+          // for the space the legend takes up with design, so starting with the original
+          // designs to get the core functionality out, and then we can tune the
+          // UI/UX in a follow-up.
```
Comment (juliepagano) [inline ui/panels-plugin/src/plugins/time-series-chart/TimeSeriesChartPanel.tsx:251]: I grabbed time with @foxbat07 on Friday to dig into this a bit more.

File: ui/core/src/utils/time-series-data.ts [hunk 5]
```diff
@@ -107,3 +107,80 @@ export function getCommonTimeScale(seriesData: Array<TimeSeriesData | undefined>
...
+      }
+    }
+  });
+
+  let averageNonNull: undefined | number = undefined;
+  if (typeof total !== 'undefined' && typeof nonNullCount !== 'undefined') {
+    averageNonNull = total > 0 && nonNullCount > 0 ? total / nonNullCount : 0;
+  }
+
+  return {
```
Comment (juliepagano) [inline ui/core/src/utils/time-series-data.ts:178]: Intentionally naming the values where excluding nulls is meaningful in case we end up wanting to support the "including nulls" cases in the future and to make it very explicit what's happening when reading the code.

File: ui/panels-plugin/src/plugins/time-series-chart/TimeSeriesChartPanel.tsx (hunk 3 above)
Comment (sjcobb) [reply ui/panels-plugin/src/plugins/time-series-chart/TimeSeriesChartPanel.tsx:238]: Sounds good, maybe let’s reconvene after we look more into the ‘time’ axis rewrite in #1134 too, that will also simplify a bit of the mapping, I’m planning to pair with Eunice on that soon to split up the TODOs

File: schemas/panels/time-series/time-series.cue [hunk 6]
```diff
@@ -17,9 +17,12 @@ import (
 	"github.com/perses/perses/schemas/common"
 )
 
+#legend_value: "AverageNonNull" | "FirstNonNull" | "LastNonNull" | "Min" | "Max" | "Total"
```
Comment (sjcobb) [inline schemas/panels/time-series/time-series.cue:20]: I prefer `calculation` to be consistent with GaugeChart and StatChart

Comment (sjcobb) [reply schemas/panels/time-series/time-series.cue:20]: Could be good to add these options as supported calculations in those other panel types too in the future so same types / utils can be reused

File: ui/core/src/utils/time-series-data.ts (hunk 5 above)
Comment (sjcobb) [reply ui/core/src/utils/time-series-data.ts:178]: We should be trying to consolidate with these utils: https://github.com/perses/perses/blob/main/ui/plugin-system/src/model/calculations.ts

They really shouldn’t be in plugin-system though ([context](https://github.com/perses/perses/pull/761#discussion_r1022875646)), I think calculation.ts utils should move to core and CalculationSelector to components

Comment (juliepagano) [reply ui/core/src/utils/time-series-data.ts:178]: Welp, I completely missed that we had those. 😓  Definitely one of the side effects of the codebase being so huge and it not always being clear where to look for something. I'll take a pass at combining them.

**Comment by juliepagano** [comment]
Converting this back to draft while I fix up the calculation stuff. Hoping to have an updated PR back up by end of day.

File: ui/core/src/model/calculations.test.ts [hunk 7]
```diff
@@ -12,126 +12,299 @@
...
 
 import { TimeSeriesValueTuple } from '@perses-dev/core';
-import { CalculationsMap, CalculationType } from './calculations';
+import { CalculationsMap, CalculationType, getCalculation, getCalculations } from './calculations';
 
 interface CalculationTestCase {
   values: TimeSeriesValueTuple[];
   calculation: CalculationType;
   expected: number | undefined | null;
 }
```
Comment (juliepagano) [inline ui/core/src/model/calculations.test.ts:22]: The diff looks a little weird because I moved the test cases around. A bunch of these existed already and I added a bunch of tests cases for edge cases (e.g. all nulls, empty array) and new calculations (min, max).

File: ui/core/src/model/calculations.ts [hunk 8]
```diff
@@ -40,11 +41,11 @@ export const CALCULATIONS_CONFIG: Readonly<Record<CalculationType, CalculationCo
     description: 'Last value',
   },
   FirstNumber: {
-    label: 'First number',
+    label: 'First #',
```
Comment (juliepagano) [inline ui/core/src/model/calculations.ts:44]: I need to check with design about how to shorten these. The word number is just too long to keep the columns small in the table.

File: ui/core/src/model/calculations.ts [hunk 9]
```diff
@@ -55,56 +56,151 @@ export const CALCULATIONS_CONFIG: Readonly<Record<CalculationType, CalculationCo
...
 
 type CalculationValue = number | null | undefined;
 
+/**
+ * Calculate a multiple values for a set of time series data.
+ *
+ * @param values - Array of time series data.
+ * @param includeCalculations - Array of calculations to include.
+ */
+export function getCalculations<IncludeCalcs extends CalculationType[]>(
```
Comment (juliepagano) [inline ui/core/src/model/calculations.ts:79]: I'm hoping this is relatively readable with how I've written and commented it, but please let me know if it's confusing. Cases where we iterate over a ton of data are probably good ones to err a little more on perf optimizations over readability, but I want to make sure it's maintainable.

File: ui/core/src/model/calculations.ts [hunk 10]
```diff
@@ -55,56 +56,151 @@ export const CALCULATIONS_CONFIG: Readonly<Record<CalculationType, CalculationCo
...
+  values: TimeSeriesValueTuple[],
+  includeCalculations: IncludeCalcs
+): Record<
+  // This extract combined with the generics above keeps the key of the returned
+  // record to *just* the specified calculations.
+  Extract<CalculationType, IncludeCalcs[number]>,
+  CalculationValue
```
Comment (juliepagano) [inline ui/core/src/model/calculations.ts:86]: In other words, if you request `['First', 'Last']`, the return type would be `Record<'First' | 'Last', CalculationValue>`

File: ui/core/src/model/calculations.ts [hunk 11]
```diff
@@ -55,56 +56,151 @@ export const CALCULATIONS_CONFIG: Readonly<Record<CalculationType, CalculationCo
...
+  // in the calculations because they are needed by multiple calculations.
+  let nonNullCount = 0;
+  let sum = 0;
+
+  // We use this large function capable of performing one or more calculations
+  // in a single iteration of the data to minimize the performance impact of
+  // generating multiple calculations for large timeseries values. This is
+  // less optimized for certain single calculations when done in isolation (e.g.
+  // `Last`), but will be more performant in the more expensive cases where
+  // multiple values are being used (e.g. table legend).
```
Comment (juliepagano) [inline ui/core/src/model/calculations.ts:103]: This is a very long way of explaining that it's better to iterate once if you're using really large data sets, which we can have with some time series.

Comment (juliepagano) [reply ui/core/src/model/calculations.ts:103]: For context, `Last` would be cheaper as a separate calc because you can just do `values[values.length - 1]` and completely avoid iterating all the values. We could put a check for that in here, but I think it would overly complicate the code for a tiny optimization (as opposed to the potentially larger optimization around calculating multiple values that do require iterating).

File: internal/api/shared/migrate/mapping.cuepart [hunk 12]
```diff
@@ -19,12 +19,30 @@ import (
     }
     // mapping table for the calculation attribute (key = grafana unit, value = perses equivalent)
     calc: {
```
Comment (juliepagano) [inline internal/api/shared/migrate/mapping.cuepart:21]: Adjusted these a bunch to try to account for all the various use cases in old & new grafana.

File: schemas/panels/time-series/mig.cuepart [hunk 13]
```diff
@@ -3,14 +3,17 @@ if #panel.type == "timeseries" || #panel.type == "graph" {
     spec: {
         legend: {
             if #panel.type == "timeseries" {
-            	  position: [
```
Comment (juliepagano) [inline schemas/panels/time-series/mig.cuepart:6]: Fixing some whitespace here to be consistent with the one below.

File: schemas/panels/time-series/mig.cuepart [hunk 14]
```diff
@@ -21,6 +24,16 @@ if #panel.type == "timeseries" || #panel.type == "graph" {
                     if #panel.legend.alignAsTable != _|_ if #panel.legend.alignAsTable { "Table" },
                     { "List" }
                 ][0]
+                values: [ for oldCalc, newCalc in #mapping.calc  
+                    // Check if the mapping field is set on the legend.
+                    if (#panel.legend[oldCalc] != _|_)  
+                        // Check if the mapping field is true. This has to be
+                        // done inside the prior check to avoid cue throwing
+                        // undefined errors when the value is not present.
```
Comment (juliepagano) [inline schemas/panels/time-series/mig.cuepart:32]: I *really* wish cue was smart enough to fail after the first conditional so I could do a single `if` with "and" logic, but sadly, it does not.

Comment (sjcobb) [reply schemas/panels/time-series/mig.cuepart:32]: @AntoineThebaud can we get some help with a few CUE questions, Julie put another one in Matrix as well

It’s probably fine to do the migration updates later, especially if it isn’t a 1:1 mapping like most of the examples in other cuepart files

Comment (juliepagano) [reply schemas/panels/time-series/mig.cuepart:32]: I _think_ I got everything working although it took me forever to figure it out because I don't know cue very well. It'll be good to get a look from Antoine as a reviewer to make sure I got it right though!

The comment about the conditionals was mostly context for reviewers (and a little bit of me whining) because this works differently than a lot of other languages. I think that's the best way to do it though based on looking at some [discussions](https://github.com/cue-lang/cue/discussions/2061#discussioncomment-4044662) in the cue github.

**Comment by juliepagano** [comment]
This should be ready to review now. Notable changes:
- A ton of refactoring of the calculation utils to be used across table legend and calculating values for stat/gauge chart. 
- Fixing some of the cuelang types and making the migrations work. @AntoineThebaud -- I mostly added you to review the cuelang changes I made. Feel free to skip the frontend code and leave reviewing that to @sjcobb and/or @saminzadeh unless you're super interested in it.

**Review by saminzadeh — APPROVED** [review]
This looks awesome - I can't wait to use it!

File: ui/core/src/model/calculations.ts (hunk 9 above)
Comment (juliepagano) [reply ui/core/src/model/calculations.ts:79]: 💯 I was debating making it a little more generic, but decided not to in this PR because it was already big enough, but could totally see us doing so the next time we need to touch this.

File: ui/core/src/model/calculations.ts (hunk 8 above)
Comment (juliepagano) [reply ui/core/src/model/calculations.ts]: Design decision from @foxbat07: use an asterisk. Pushed up a change for that.

**Review by sjcobb — APPROVED** [review]
🚀 🚀
